pkg/orchestrator/evaluation/inmemory_types.go

```diff
@@ -0,0 +1,174 @@
+package evaluation
+
+import (
+	"container/heap"
+	"time"
+
+	"github.com/bacalhau-project/bacalhau/pkg/model"
+)
+
+// evalWrapper satisfies the ScheduledTask interface
+type evalWrapper struct {
+	eval *model.Evaluation
+}
+
+func (d *evalWrapper) GetData() *model.Evaluation {
+	return d.eval
+}
+
+func (d *evalWrapper) GetID() string {
+	return d.eval.ID
+}
+
+func (d *evalWrapper) GetWaitUntil() time.Time {
+	return d.eval.WaitUntil
+}
+
+// inflightEval tracks an unacknowledged evaluation along with the visibility timeout timer
+type inflightEval struct {
+	Eval            *model.Evaluation
+	ReceiptHandle   string
+	VisibilityTimer *time.Timer
+}
+
+// BrokerStats returns all the stats about the broker
+type BrokerStats struct {
+	TotalReady      int
+	TotalInflight   int
+	TotalPending    int
+	TotalWaiting    int
+	TotalCancelable int
+	DelayedEvals    map[string]*model.Evaluation
+	ByScheduler     map[string]*SchedulerStats
+}
+
+// IsEmpty returns true if the stats are zero
+func (s *BrokerStats) IsEmpty() bool {
+	if len(s.DelayedEvals) > 0 {
+		return false
+	}
+	for _, v := range s.ByScheduler {
+		if !v.IsEmpty() {
+			return false
+		}
+	}
+	return s.TotalReady == 0 &&
+		s.TotalInflight == 0 &&
+		s.TotalPending == 0 &&
+		s.TotalWaiting == 0 &&
+		s.TotalCancelable == 0
+}
+
+// SchedulerStats returns the stats per scheduler
+type SchedulerStats struct {
+	Ready    int
+	Inflight int
+}
+
+// IsEmpty returns true if the scheduler stats are zero
+func (s *SchedulerStats) IsEmpty() bool {
+	return s.Ready == 0 && s.Inflight == 0
+}
+
+// ReadyEvaluations is a list of ready Evaluations across multiple jobs. We
+// implement the container/heap interface so that this is a priority queue.
+type ReadyEvaluations []*model.Evaluation
+
+// Len is for the sorting interface
+func (r ReadyEvaluations) Len() int {
+	return len(r)
+}
+
+// Less is for the sorting interface. We flip the check
+// so that the "min" in the min-heap is the element with the
+// highest priority
+func (r ReadyEvaluations) Less(i, j int) bool {
+	if r[i].JobID != r[j].JobID && r[i].Priority != r[j].Priority {
+		return !(r[i].Priority < r[j].Priority)
+	}
+	return r[i].CreateIndex < r[j].CreateIndex
+}
+
+// Swap is for the sorting interface
+func (r ReadyEvaluations) Swap(i, j int) {
+	r[i], r[j] = r[j], r[i]
+}
+
+// Push is used to add a new evaluation to the slice
+func (r *ReadyEvaluations) Push(e interface{}) {
+	*r = append(*r, e.(*model.Evaluation))
+}
+
+// Pop is used to remove an evaluation from the slice
+func (r *ReadyEvaluations) Pop() interface{} {
+	n := len(*r)
+	e := (*r)[n-1]
+	(*r)[n-1] = nil
```

Comment:
Not a comment, more a query, is there much value in setting the last element to nil if we're going to replace the slice (without this element in it) in the next line? Does it help with gc?

Reply:
it doesn't look like it adds value. removing it in next revision